functions/github/webhooks.js

```diff
@@ -0,0 +1,154 @@
+const {
+  getReleaseIdFromBranch,
+  updateCheckRunStatus,
+} = require("../database/database.js");
+const {
+  syncReleaseState,
+} = require("../handlers/handlers.js");
+const {
+  verifySignature,
+} = require("./github.js");
+const {Octokit} = require("@octokit/rest");
+const {error, log} = require("firebase-functions/logger");
+const {defineSecret} = require("firebase-functions/params");
+const GITHUB_WEBHOOK_SECRET = defineSecret("GITHUB_WEBHOOK_SECRET");
+const GITHUB_TOKEN = defineSecret("GITHUB_TOKEN");
+
+/**
+  * Handles a GitHub check run event. This function is called when a check run
+  * is created, completed, or rerequested. It updates the release state if
+  * necessary.
+  *
+  * @param {Object} req - The request object.
+  * @param {Object} res - The response object.
+  * @return {Promise} A promise that resolves when the request is complete.
+  */
+async function githubWebhook(req, res) {
+  if (!verifySignature(req, GITHUB_WEBHOOK_SECRET.value())) {
+    return res.status(401).send("Unauthorized");
+  }
+
+  const payload = req.body;
+  const eventType = req.headers["x-github-event"];
+
+  log("Received GitHub webhook", {payload: payload, eventType: eventType});
+
+  try {
+    if (eventType === "check_run") {
+      await handleCheckRunEvent(payload);
+    } else if (eventType === "pull_request") {
+      await handlePullRequestEvent(payload);
+    }
+  } catch (err) {
+    error("Failed to handle GitHub webhook",
+        {
+          error: err.message,
+          payload: payload,
+          eventType: eventType,
+        },
+    );
+    return res.status(500).send("Internal Server Error");
+  }
+
+  return res.status(200).send("OK");
+}
+
+/**
+  * Handles a GitHub pull request event.
+  *
+  * Updates the release state if the the event implies that the release state
+  * has changed, and needs to be synchronized. As of now, those two events are
+  * opens, and synchronizes (pushes to the pull request branch).
+  *
+  * @param {Object} payload The payload from the GitHub webhook.
+  * @return {Promise<void>}
+  */
+async function handlePullRequestEvent(payload) {
+  if (payload.action === "opened" || payload.action == "synchronize") {
+    const pullRequest = payload.pull_request;
+
+    if (pullRequest && pullRequest.head) {
```

Comment:
When would this evaluate to false?

Reply:
Oh, you're right, this is authenticated so there's no need to verify the payload this much- useless if statement. Will change